libcli.c

```diff
@@ -1966,11 +1990,1146 @@ void cli_set_idle_timeout_callback(struct cli_def *cli, unsigned int seconds, in
+  if (comphelp) {
+    int idx;
+
+    for (idx = 0; idx < comphelp->num_entries; idx++) free_z(comphelp->entries[idx]);
+    free_z(comphelp->entries);
+  }
+}
+
+static int cli_int_locate_command(struct cli_def *cli, struct cli_command *commands, int command_type, int start_word,
+                                  struct cli_pipeline_stage *stage) {
+  struct cli_command *c, *again_config = NULL, *again_any = NULL;
+  int c_words = stage->num_words;
+
+  for (c = commands; c; c = c->next) {
+    if (c->command_type != command_type) continue;
+    if (cli->privilege < c->privilege) continue;
+
+    if (strncasecmp(c->command, stage->words[start_word], c->unique_len)) continue;
+
+    if (strncasecmp(c->command, stage->words[start_word], strlen(stage->words[start_word]))) continue;
+
+  AGAIN:
+    if (c->mode == cli->mode || (c->mode == MODE_ANY && again_any != NULL)) {
+      int rc = CLI_OK;
+
+      // Found a word!
+      if (!c->children) {
+        // Last word
+        if (!c->callback && !c->filter) {
+          cli_error(cli, "No callback for \"%s\"", cli_command_name(cli, c));
+          return CLI_ERROR;
+        }
+      } else {
+        if (start_word == c_words - 1) {
+          if (c->callback) goto CORRECT_CHECKS;
+
+          cli_error(cli, "Incomplete command");
+          return CLI_ERROR;
+        }
+        rc = cli_int_locate_command(cli, c->children, command_type, start_word + 1, stage);
+        if (rc == CLI_ERROR_ARG) {
+          if (c->callback) {
+            rc = CLI_OK;
+            goto CORRECT_CHECKS;
+          } else {
+            cli_error(cli, "Invalid %s \"%s\"", commands->parent ? "argument" : "command", stage->words[start_word]);
+          }
+        }
+        return rc;
+      }
+
+      if (!c->callback && !c->filter) {
+        cli_error(cli, "Internal server error processing \"%s\"", cli_command_name(cli, c));
+        return CLI_ERROR;
+      }
+
+    CORRECT_CHECKS:
+
+      if (rc == CLI_OK) {
+
+        stage->command = c;
+        stage->first_unmatched = start_word + 1;
+        stage->first_optarg = stage->first_unmatched;
+        cli_int_parse_optargs(cli, stage, c, '\0', NULL);
+        rc = stage->status;
+      }
+      return rc;
+    } else if (cli->mode > MODE_CONFIG && c->mode == MODE_CONFIG) {
+      // command matched but from another mode,
+      // remember it if we fail to find correct command
+      again_config = c;
+    } else if (c->mode == MODE_ANY) {
+      // command matched but for any mode,
+      // remember it if we fail to find correct command
+      again_any = c;
+    }
+  }
+
+  // drop out of config submode if we have matched command on MODE_CONFIG
+  if (again_config) {
+    c = again_config;
+    goto AGAIN;
+  }
+  if (again_any) {
+    c = again_any;
+    goto AGAIN;
+  }
+
+  if (start_word == 0)
+    cli_error(cli, "Invalid %s \"%s\"", commands->parent ? "argument" : "command", stage->words[start_word]);
```

Comment:
This has the potential for a null pointer deref. `commands` isn't guaranteed here.

Reply:
Not as used.  The callers won't pass a null to cli_int_locate_command(), nor will it pass a null to itself when called recursively.  However I wonder if this ought to be  'commands' should be 'c' instead.  I modeled cli_int_locate_command greatly on cli_find_command, but altered to use the pipeline constructs.